Create MultiBrush brush

PunkPun opened the pull request:
Closes #10655

Showcase of multibrush: https://www.youtube.com/watch?v=0xSitFxY6zc

This tool allows players to apply multibrush in the map editor. What is multibrush? It's defined in tileset definition and allows to combine actors and tiles under one umbrella. It's used for randomized painting.

In the future we will want to add more multibrushes as well as create an ingame UI for creating user defined multibrushes.

PunkPun committed:
Fix multibrush ignoring bounds

ashleynewson commented on OpenRA.Mods.Common/EditorBrushes/EditorMultiBrushBrush.cs:
```diff
@@ -0,0 +1,169 @@
+
+namespace OpenRA.Mods.Common.Widgets
+{
+	public sealed class EditorMultiBrushBrush : IEditorBrush
```
I did not foresee this problem when I named it MultiBrush. :sob: 

Do we want "MultiBrush" to be the user-facing name for the tool? After all, it's for a specific usage of MultiBrush.

PunkPun replied:
We could call it PaintBrush

PunkPun replied:
named the user facing value to Paint Brush

ashleynewson replied:
Should we update the code to reflect?

PunkPun replied:
renamed to EditorPaintBrush

ashleynewson commented on OpenRA.Mods.Common/MapGenerator/MultiBrush.cs:
```diff
@@ -615,6 +615,71 @@ void PaintActors(Map map, List<ActorPlan> actorPlans, CPos paintAt)
 			}
 		}
 
+		public void Paint(MultiBrush multiBrush,
```
- Clear the cached shape with `multiBrush.shape = null`.
- Lift the repeated code, either using nested methods or using something like a `willCopyTiles` boolean and doing it after the switch statement.
- Keep it next to the other `Paint` overload.
- Add documentation above the method. :pleading_face: 
- There is the existing `MergeFrom` method, which this is similar to (but inadequate by itself). Perhaps there's an opportunity for deduplication/refactoring.

ashleynewson commented on OpenRA.Mods.Common/MapGenerator/MultiBrush.cs:
```diff
@@ -687,7 +752,11 @@ Replaceability ReserveShape(CPos paintAt, IEnumerable<CVec> shape, Replaceabilit
 					var cpos = paintAt + cvec;
 					if (!replace.Contains(cpos))
-						continue;
+					{
```
The implication of `!replace.Contains(cpos)` is (or was) that `cpos` is outside of the map's CellLayer bounds. (Admittedly, there was no check that `replace` has the same shape as the map, and I think that your change means that's not a guarantee for the other overload?)

I think the intent here was so that large MultiBrushes are allowed to extend outside of the map without having any contract constraints placed on them.

Whilst this change will now disallow that, I guess I'm not opposed to it, as I can imagine the existing solution wasn't fair to all map edges. :+1:

ashleynewson commented on OpenRA.Mods.Common/MapGenerator/MultiBrush.cs:
```diff
@@ -745,6 +814,140 @@ Replaceability ReserveShape(CPos paintAt, IEnumerable<CVec> shape, Replaceabilit
 			}
 		}
 
+		public static MultiBrush PaintAreaBrush(
```
We should probably avoid the method implementation duplication. I think:

- Make this method the one with the actual implementation.
- Redefine the existing PaintArea in terms of this one (it can create the composed brush and then just paint it).

If we have this new PaintArea that creates a MultiBrush instead of painting directly to a map, we should be able to refactor it a bit. For example, in principle, it wouldn't need the map as an argument. That can be done in a follow-up though. I'm happy for an issue to be assigned to me to do that cleanup if you want to avoid digging in too deep.

PunkPun replied:
It's very hard to detach into parts. I have seperated out a few functions

ashleynewson commented on OpenRA.Mods.Common/MapGenerator/MultiBrush.cs:
```diff
@@ -754,7 +957,8 @@ Replaceability ReserveShape(CPos paintAt, IEnumerable<CVec> shape, Replaceabilit
 			MersenneTwister random,
-			PlayerReference defaultActorOwner = null)
+			PlayerReference defaultActorOwner = null,
+			CellRegion region = null)
```
What is the value of passing in a region here? Is there some bug that appears without it?

PunkPun replied:
We don't want the brush size to be dynamic. The original function was meant for tiling path

ashleynewson commented on OpenRA.Mods.Common/Traits/World/MultiBrushTool.cs:
```diff
@@ -0,0 +1,184 @@
+					MultiBrushes[collection.Key] = brushes = [];
+
+				MultiBrushes[collection.Key] = brushes.AddRange(MultiBrush.LoadCollection(World.Map, collection.Key));
+			}
```
```suggestion
			foreach (var collection in templatedTerrainInfo.MultiBrushCollections.Keys)
				MultiBrushes[collection] = MultiBrush.LoadCollection(World.Map, collection);
```

(Or some LINQ)

ashleynewson commented on OpenRA.Mods.Common/Traits/World/MultiBrushTool.cs:
```diff
@@ -0,0 +1,184 @@
+		public WorldRenderer WorldRenderer;
+		public Dictionary<string, ImmutableArray<MultiBrush>> MultiBrushes = [];
+
+		public (string Name, ImmutableArray<MultiBrush> Brushes) SelectedMultiBrush;
```
It might be more accurate to rename:

`MultiBrushes` -> `MultiBrushCollections`
`SelectedMultiBrush` -> `SelectedMultiBrushCollection`

ashleynewson reviewed (commented):
I think the user should be able to continuously draw with the brush (mouse  down and drag) to create arbitrary shapes, with the filling be applied seamlessly throughout. Perhaps this could be done by staging the edit until its confirmed. This would also allow for some degree of erasing whilst the final shape is being finalized by the user. At the moment, trying to paint larger, awkward shapes by overlapping squares/circles can result in broken templates or holes without actors.

The existing UI has trained me to expect right-click to go to the selection brush, which also doesn't happen here. Perhaps for both the Path Tiling tool and the MultiBrush tool we should exit from the tool (change out the panel) when using right click.

Being able to floodfill based on tiles or actors might also be a cheap interim alternative to a fully fledged area drawing interface. (As with the shift-key in the tile editor.)

Alternatively, I wonder if being able to fill a designated marker tile color would work well. That could give some drawing and symmetry tooling for free whilst unifying the workflows.

We should exclude MultiBrushes which have a segment definition, as this doesn't make sense to use in this tool. Any MultiBrush collections which contain only segmented MultiBrushes should be excluded from the dropdown.

PunkPun committed:
Add some TS multi-brushes

PunkPun commented:
Updated, now the brushes work in TS as well

> I think the user should be able to continuously draw with the brush (mouse down and drag) to create arbitrary shapes, with the filling be applied seamlessly throughout. Perhaps this could be done by staging the edit until its confirmed. This would also allow for some degree of erasing whilst the final shape is being finalized by the user. At the moment, trying to paint larger, awkward shapes by overlapping squares/circles can result in broken templates or holes without actors.

Added drawing

> Being able to floodfill based on tiles or actors might also be a cheap interim alternative to a fully fledged area drawing interface. (As with the shift-key in the tile editor.)

Added floodfill

> We should exclude MultiBrushes which have a segment definition, as this doesn't make sense to use in this tool. Any MultiBrush collections which contain only segmented MultiBrushes should be excluded from the dropdown.

I just added a way to exclde brushes

ashleynewson commented:
The drawing feels more natural now and the flood fill seems quite powerful. I still think marker tile filling could be the most ergonomic option.

A few bugs I noticed:

I came across a crash when drag-drawing a shape that's larger than the map dimensions and crosses into the negative X or Y coordinate space. I recorded a reproduction video (the recording goes black at the time of crash).

<details>
<summary>Video and stack traces</summary>

https://github.com/user-attachments/assets/ebcf0d83-9adc-42c9-9ceb-474bbf88812e

RA:
```
Exception of type `System.IndexOutOfRangeException`: Index was outside the bounds of the array.
   at OpenRA.CellLayer`1.Index(MPos uv) in /home/ashley/devel/OpenRA/OpenRA.Game/Map/CellLayer.cs:line 75
   at OpenRA.Mods.Common.Graphics.EditorCellLayerAnnotationRenderable.Render(WorldRenderer wr) in /home/ashley/devel/OpenRA/OpenRA.Mods.Common/Graphics/EditorCellLayerAnnotationRenderable.cs:line 60
   at OpenRA.Graphics.WorldRenderer.DrawAnnotations() in /home/ashley/devel/OpenRA/OpenRA.Game/Graphics/WorldRenderer.cs:line 336
   at OpenRA.Game.RenderTick() in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 738
   at OpenRA.Game.Loop() in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 878
   at OpenRA.Game.Run() in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 910
   at OpenRA.Game.InitializeAndRun(String[] args) in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 339
   at OpenRA.Launcher.Program.Main(String[] args) in /home/ashley/devel/OpenRA/OpenRA.Launcher/Program.cs:line 46
```

TS:
```
Exception of type `System.IndexOutOfRangeException`: Index was outside the bounds of the array.
   at OpenRA.Mods.Common.Graphics.EditorCellLayerAnnotationRenderable.Render(WorldRenderer wr) in /home/ashley/devel/OpenRA/OpenRA.Mods.Common/Graphics/EditorCellLayerAnnotationRenderable.cs:line 44
   at OpenRA.Graphics.WorldRenderer.DrawAnnotations() in /home/ashley/devel/OpenRA/OpenRA.Game/Graphics/WorldRenderer.cs:line 336
   at OpenRA.Game.RenderTick() in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 738
   at OpenRA.Game.Loop() in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 878
   at OpenRA.Game.Run() in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 910
   at OpenRA.Game.InitializeAndRun(String[] args) in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 339
   at OpenRA.Launcher.Program.Main(String[] args) in /home/ashley/devel/OpenRA/OpenRA.Launcher/Program.cs:line 46

```
</details>

Releasing the mouse button doesn't get handled if it happens on top of the editor panel. (Or minimap, but that has other issues!)

ashleynewson commented on OpenRA.Mods.Common/MapGenerator/CellLayerUtils.cs:
```diff
@@ -392,6 +444,36 @@ public static Rectangle CellBounds(Map map)
 			return CellBounds(map.MapSize, map.Grid.Type);
 		}
 
+		/// <summary>Returns a cell region containing cell coords.</summary>
```
It returns a CellLayer, not a "cell region"?

ashleynewson commented on OpenRA.Mods.Common/MapGenerator/CellLayerUtils.cs:
```diff
@@ -236,55 +303,40 @@ public static CPos[][] FromMatrixPoints<T>(
 		/// </summary>
 		public static void OverCircle<T>(
 			CellLayer<T> cellLayer,
+			CellCoordsRegion region,
```
The region can be inferred from the CellLayer in most cases.

We could have two overloads: one that uses only the CellLayer, and one which uses a gridType and region. (The CellLayer variant could be defined trivially in terms of the other one.)

PunkPun replied:
In isometric they are fundamentally different, although the space matches if CellRegion was instantiated with CPos

PunkPun replied:
Added the overload

PunkPun commented:
Fixed the crash, and added additional toggles and bug fixes

> Releasing the mouse button doesn't get handled if it happens on top of the editor panel.

That's a bug with all brushes. I've coded it in such a way it would handle such a case nicelly

> I still think marker tile filling could be the most ergonomic option.

I'm not sure if it is. It would need quite a bit of experimenting to get right. I know the current setup isn't perfect, but I fear the easy alternative is worse

PunkPun commented:
Added a brush for placing single instances, added sparcity, and more bug fixes.

It's an incredibly powerful tool. Code can definitelly still be improved, however it is now more in a state of ready to be merged.

ashleynewson commented:
I get a crash on entering the map editor (any mod, new or loaded map).

(commit aa651cf1ab23649c6436ee5cba6c494a503502d3)

<details>
<summary>Stack trace</summary>

```
Exception of type `System.ArgumentException`: cells must not be null or empty. (Parameter 'cells')
   at OpenRA.CellCoordsRegion.BoundingRegion(MapGridType shape, IReadOnlyCollection`1 cells) in /home/ashley/devel/OpenRA/OpenRA.Game/Map/CellCoordsRegion.cs:line 119
   at OpenRA.Mods.Common.Traits.EditorActorLayer.UpdateNeighbours(EditorActorPreview[] previews) in /home/ashley/devel/OpenRA/OpenRA.Mods.Common/Traits/World/EditorActorLayer.cs:line 310
   at OpenRA.Mods.Common.Traits.EditorActorLayer.AddRange(List`1 references, String[] names) in /home/ashley/devel/OpenRA/OpenRA.Mods.Common/Traits/World/EditorActorLayer.cs:line 194
   at OpenRA.Mods.Common.Traits.EditorActorLayer.WorldLoaded(World world, WorldRenderer wr) in /home/ashley/devel/OpenRA/OpenRA.Mods.Common/Traits/World/EditorActorLayer.cs:line 102
   at OpenRA.World.LoadComplete(WorldRenderer wr) in /home/ashley/devel/OpenRA/OpenRA.Game/World.cs:line 294
   at OpenRA.Game.StartGame(Map map, WorldType type) in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 229
   at OpenRA.Game.StartGame(String uid, WorldType type) in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 191
   at OpenRA.Game.LoadEditor(String uid) in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 536
   at OpenRA.Mods.Common.Widgets.Logic.MainMenuLogic.LoadMapIntoEditor(String uid) in /home/ashley/devel/OpenRA/OpenRA.Mods.Common/Widgets/Logic/MainMenuLogic.cs:line 370
   at OpenRA.Mods.Common.Widgets.Logic.MainMenuLogic.<>c__DisplayClass17_0.<.ctor>b__17(String uid) in /home/ashley/devel/OpenRA/OpenRA.Mods.Common/Widgets/Logic/MainMenuLogic.cs:line 193
   at OpenRA.Mods.Common.Widgets.Logic.MapChooserLogic.<>c__DisplayClass56_1.<EnumerateMaps>g__DblClick|4() in /home/ashley/devel/OpenRA/OpenRA.Mods.Common/Widgets/Logic/MapChooserLogic.cs:line 538
   at OpenRA.Mods.Common.Widgets.ButtonWidget.HandleMouseInput(MouseInput mi) in /home/ashley/devel/OpenRA/OpenRA.Mods.Common/Widgets/ButtonWidget.cs:line 177
   at OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi) in /home/ashley/devel/OpenRA/OpenRA.Game/Widgets/Widget.cs:line 436
   at OpenRA.Widgets.Ui.HandleInput(MouseInput mi) in /home/ashley/devel/OpenRA/OpenRA.Game/Widgets/Widget.cs:line 107
   at OpenRA.DefaultInputHandler.<>c__DisplayClass5_0.<OnMouseInput>b__0() in /home/ashley/devel/OpenRA/OpenRA.Game/Input/InputHandler.cs:line 50
   at OpenRA.Sync.RunUnsynced[T](Boolean checkSyncHash, World world, Func`1 fn) in /home/ashley/devel/OpenRA/OpenRA.Game/Sync.cs:line 192
   at OpenRA.DefaultInputHandler.OnMouseInput(MouseInput input) in /home/ashley/devel/OpenRA/OpenRA.Game/Input/InputHandler.cs:line 50
   at OpenRA.Platforms.Default.Sdl2Input.PumpInput(Sdl2PlatformWindow device, IInputHandler inputHandler, Nullable`1 lockedMousePosition) in /home/ashley/devel/OpenRA/OpenRA.Platforms.Default/Sdl2Input.cs:line 144
   at OpenRA.Platforms.Default.Sdl2PlatformWindow.PumpInput(IInputHandler inputHandler) in /home/ashley/devel/OpenRA/OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:line 498
   at OpenRA.Renderer.EndFrame(IInputHandler inputHandler) in /home/ashley/devel/OpenRA/OpenRA.Game/Renderer.cs:line 326
   at OpenRA.Game.RenderTick() in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 755
   at OpenRA.Game.Loop() in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 878
   at OpenRA.Game.Run() in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 910
   at OpenRA.Game.InitializeAndRun(String[] args) in /home/ashley/devel/OpenRA/OpenRA.Game/Game.cs:line 339
   at OpenRA.Launcher.Program.Main(String[] args) in /home/ashley/devel/OpenRA/OpenRA.Launcher/Program.cs:line 46
```

</details>

PunkPun commented:
ah, makes sense. There aren't any actors on the map

ashleynewson commented:
> I get a crash on entering the map editor (any mod, new or loaded map).
> 
> (commit [aa651cf](https://github.com/OpenRA/OpenRA/commit/aa651cf1ab23649c6436ee5cba6c494a503502d3))

Looks like it's the changes from #22009 ?

PunkPun commented:
fixed

ashleynewson commented on OpenRA.Mods.Common/EditorBrushes/EditorMultiBrushBrush.cs:
```diff
@@ -0,0 +1,135 @@
+
+		public bool HandleKeyboardInput(KeyInput ki)
+		{
+			if (ki.Key == Keycode.LCTRL || ki.Key == Keycode.RCTRL)
```
The floodfill for terrain tiles uses the shift key, but this uses the ctrl key. They should probably be the same?

PunkPun replied:
In my other work I have made shift key draw lines, while moving flood fill to Ctrl. I haven't had time to add line dragging to this too yet

But sure, I can change it to shift in the meantime

ashleynewson commented on OpenRA.Game/Support/MersenneTwister.cs:
```diff
@@ -99,23 +98,24 @@ public float NextFloat()
-			int i;
 			ulong acc = 0;
-			for (i = 0; i < weights.Count; i++)
+			for (i = 0; i < weights.Length; i++)
```
Partly my fault to begin with, but both uses of `i` could become loop-scoped `var i = 0` instead.

PunkPun replied:
Tbf, I didn't expect getting an iterator from IReadOnlyList so have such a large overhead. I was really suprised when this function was marked as having a ton of allocations

ashleynewson commented on OpenRA.Mods.Common/EditorBrushes/EditorBlit.cs:
```diff
@@ -45,6 +46,7 @@ public sealed class EditorBlit
 		readonly CPos blitPosition;
 		readonly Map map;
 		readonly bool respectBounds;
+		readonly bool stickToGround;
```
Perhaps we should have stickToGround as part of MapBlitFilters? (Height)

PunkPun replied:
Interesting idea. Would have to think about it more

ashleynewson commented on OpenRA.Mods.Common/EditorBrushes/EditorBlit.cs:
```diff
@@ -211,42 +222,63 @@ public static IEnumerable<IRenderable> PreviewBlitSource(
+						continue;
+
+					byte height;
+					if (stickToGround)
+					{
+						if (!mapHeight.TryGetValue(cPos, out height))
+							height = byte.MinValue;
+
+						if (filters.HasFlag(MapBlitFilters.Terrain))
+							height += map.Rules.TerrainInfo.GetTerrainInfo(tile.TerrainTile).Height;
+					}
+					else
+					{
+						if (filters.HasFlag(MapBlitFilters.Terrain))
+							height = tile.Height;
+						else if (!mapHeight.TryGetValue(cPos, out height))
+							height = byte.MinValue;
+					}
```
Arguably slightly simpler, albeit perhaps slightly less optimized:

```suggestion
					if (!mapHeight.TryGetValue(cPos, out var height))
						height = byte.MinValue;

					if (filters.HasFlag(MapBlitFilters.Terrain))
					{
						if (stickToGround)
							height += map.Rules.TerrainInfo.GetTerrainInfo(tile.TerrainTile).Height;
						else
							height = tile.Height;
					}
```

PunkPun replied:
I think the current version better represents intent

PunkPun replied:
Adjusted to even better represent intent

ashleynewson commented on OpenRA.Mods.Common/EditorBrushes/EditorDefaultBrush.cs:
```diff
@@ -354,10 +357,10 @@ sealed class DeleteAreaAction : IEditorAction
 		readonly Map map;
 
-		public DeleteAreaAction(Map map, MapBlitFilters blitFilters, CellRegion area, IResourceLayer resourceLayer, EditorActorLayer editorActorLayer)
+		public DeleteAreaAction(Map map, MapBlitFilters blitFilters, CellCoordsRegion area, IResourceLayer resourceLayer, EditorActorLayer editorActorLayer)
```
I guess this could be made to use EditorBlit to deduplicate a bunch of logic.

(Perhaps that can be filed as an issue for follow-up to reduce the growth of this individual PR.)

PunkPun replied:
Makes sense

ashleynewson commented on OpenRA.Mods.Common/Graphics/EditorSelectionAnnotationRenderable.cs:
```diff
@@ -20,11 +20,11 @@ namespace OpenRA.Mods.Common.Graphics
 	public class EditorSelectionAnnotationRenderable : IRenderable, IFinalizedRenderable
```
The duplicated code between EditorSelectionAnnotationRenderable and EditorCellLayerAnnotationRenderable could be lifted into a shared method.

Alternatively, should they become a single class, perhaps using an `IEnumerable<CPos>` instead of a `CellLayer`/`CellCoordsRegion`?

PunkPun replied:
I'm not sure what the best approach is. Ideally we should be reusing renderables

PunkPun replied:
Implemented as IEnumerable, that cleaned quite a few things up

ashleynewson commented on OpenRA.Mods.Common/MapGenerator/CellLayerUtils.cs:
```diff
@@ -685,5 +773,40 @@ public static CellLayer<T> Create<T>(Map map, Func<CPos, T> func)
 		}
+
+		/// <summary>
+		/// Apply a mask difference to the output layer, applying only the true values from the target mask.
```
The documentation might need more elaboration. It's not clear how to use this method without looking at the implementation, and the parameter/variable names are confusing. (inconsistent names: target/new/true vs old/old/false, replacable (what's replaced?).)

PunkPun replied:
updated

ashleynewson commented on mods/common/fluent/common.ftl:
```diff
@@ -1114,3 +1117,26 @@ notification-tiling-path-started = Started tiling path
+label-flood-behavior-all-terrain = All Terrain
+label-flood-behavior-template-types = Template Types
+label-flood-behavior-terrain-types = Terrain Types
+label-flood-behavior-target-types = Target Types
```
Does "Target Types" make sense to most people? (It's not clear to me.)

PunkPun replied:
Could call it "terrain target type". It's something that would be more obvious to modders, as we really don't teach the concept to players

ashleynewson commented on OpenRA.Mods.Common/Widgets/Logic/Editor/MultiBrushToolLogic.cs:
```diff
@@ -0,0 +1,245 @@
+			var ownerContainer = widget.Get<ContainerWidget>("ACTOR_OWNER");
+			ownerContainer.IsVisible = () => tool.CurrentReplaceability.HasFlag(MultiBrush.Replaceability.Actor);
+
+			var ownersDropDown = ownerContainer.Get<DropDownButtonWidget>("DROPDOWN");
```
The owners list doesn't always appear to be up-to-date. This isn't just an issue for the Paint Brush tool, but also the actors tab. This appears to be a regression, possibly with the editor actor layer optimizations?

Repoduction:
1. Load a blank map,
1. Experimental RMG tool,
1. Generate,
1. Paint Brush tool,
1. Owners dropdown does not have recently added multiplayer spawns.

PunkPun replied:
Yeah, it's a regression from there. Fixed

ashleynewson commented on OpenRA.Mods.Common/Traits/World/MultiBrushTool.cs:
```diff
@@ -0,0 +1,941 @@
+			lastMousePosition = pos;
+		}
+
+		public void AddMousePosition(CPos pos)
```
There is a bias with the current drag drawing approach where it paints the MultiBrushes incrementally as the mouse moves, rather than all at once when the drag has completed. When dragging the mouse slowly, you'll only end up with small MultiBrushes being added incrementally, because only small MultiBrushes will fit in the incremental spaces added with each mouse movement.

This can be seen, for example, when trying to place civilian buildings using a large brush. You'll only get the larger 2x2 buildings on the initial mouse down, and then only small buildings if you slowly drag.

PunkPun replied:
Yeah, I could add a disclaimer in code. The dragging looks much, much better with this transferring approach, and you can remove the bias by right clicking.

There's also bias with the Sparsity slider, and in general with painting. There's a large preference to small actors

ashleynewson replied:
Ah, I didn't realise that right-clicking while drawing would work. I guess that solves the problem, but it isn't an intuitive thing to try.

Perhaps a wiki article would be the cheap way to deal with that, or alternatively somehow integrate that hint into the tool itself.

PunkPun replied:
In many ways I feel the current way it is controlled is less than ideal. However it's decent and I want to keep itterations for a follow up

ashleynewson commented:
A bunch of resolved comments aren't reflected by the latest revision. Did you mean to upload a new version of the code yet?

PunkPun commented:
It appears I never pushed it

PunkPun commented:
Some more refactoring. Now brushes work with subcell actors, TS contains infantry MultiBrushes to test out

ashleynewson commented on OpenRA.Mods.Common/MapGenerator/MultiBrush.cs:
```diff
@@ -272,19 +273,77 @@ public sealed class MultiBrush
 			/// <summary>Area cannot be replaced by a tile or obstructing actor.</summary>
 			None = 0,
 
-			/// <summary>Area must be replaced by a different tile, and may optionally be given an actor.</summary>
```
If this is becoming a bitwise flag enum, these should probably be phased as "can be replaced by..." rather than "must be replaced by...".

PunkPun replied:
I've remove the comments, it should be self-explanatory

ashleynewson commented on OpenRA.Mods.Common/MapGenerator/MultiBrush.cs:
```diff
@@ -272,19 +273,77 @@ public sealed class MultiBrush
+
+		/// <summary>
+		/// Generate a random set of sub-cell bits based on the given <paramref name="frequency"/>.
+		/// Frequency is the chance (0-100) of each sub-cell being included.
```
Perhaps we should use 0-1000. 1000 is the typical fractional denominator used in the map generator logic as well as the default used for MultiBrush weights.

PunkPun replied:
sure

ashleynewson commented on OpenRA.Mods.Common/MapGenerator/Terraformer.cs:
```diff
@@ -936,7 +936,11 @@ public void PaintTiling(
 			MersenneTwister random,
 			MultiBrush brush)
 		{
-			brush.Paint(Map, ActorPlans, CPos.Zero, MultiBrush.Replaceability.Any, random);
```
It seems like this is a common enough pattern that we'd still want a method that applies both tiles and actors.

Or, at the very least, we should do away with the need to check first whether it has tiles or actors.

PunkPun replied:
It appears is just a few places, though sure

ashleynewson commented on OpenRA.Mods.Common/Traits/World/MultiBrushTool.cs:
```diff
@@ -0,0 +1,954 @@
+		public ShapeType CurrentShape { get; private set; } = ShapeType.Square;
+		public ShapeType CurrentFunctionalBrushType => IsFloodFill ? ShapeType.Flood : CurrentShape;
+		public Size CurrentSize { get; private set; }
+		public int CurrentSparsity { get; private set; } = 0;
```
Would "density" be a more natural concept than "sparsity"?

PunkPun replied:
"density" is the inverse. I specifically chose sparcity as my logic considers 0 to be fully filled

PunkPun commented:
Showcase of multibrush https://www.youtube.com/watch?v=0xSitFxY6zc

PunkPun committed:
Add EditorPaintBrush

PunkPun commented:
Dependencies have been merged, rebased

PunkPun commented:
rebased
